### ISSUE TITLE
Updating shared UI state on change of dynamic segment

### ISSUE THREAD

**Issue by tacomanator** [remix-run/react-router#629]
**tl;dr:** How can I update a parent's state, from a child component, when only a dynamic segment of the route changes? The parent maintains state for UI that is shared across all pages of the app, and gives children a handler to update it. Normally the children call the handler in componentDidMount, but this is not called for dynamic segments. Using the handler from componentWillReceiveProps creates an infinite loop.

A `Navigation` component is present on every page. Some of the links are provided by the pages themselves. Therefore I keep the list of links in `App.state`, and pass a handler to the pages so they can update it:

```
App = React.createClass({

  updateLinks: function (links) {
    this.setState({'links': links});
  },

  render: function () {
    <Navigation links={this.state.links} />
    <div>
      <div>
        <RouteHandler updateLinks={this.updateLinks} />
      </div>
    </div>
  }

});
```

The routes are like this:

```
<Route name="app" path="/" handler={App}>
  <Route name="form" path="/form/:id/?:action?" handler={Form} />
</Route>
```

As the user updates from `/form/123/view` to `/form/123/edit`, the `Form` component should be able to use `updateLinks` to send new navigation to `App`. However, if I try to use the handler via `componentWillReceiveProps`, it causes the parent to update which again fires `componentWillReceiveProps` and so it will loop infinitely.

I could move the `Navigation` component into each page, but that would result in duplicating other layout code (the extra divs) in each page, and this is what I'm trying to avoid.

Any suggestions on how to resolve or a better way?

**Comment by ivan9074** [comment]
Hi @tacomanator @gaearon  could you explain to me how to implement this? currently i am doing as follows:

	<Route name="app" path="/" handler={Main}>
		<Route name="losts" path="losts" handler={Losts} />
		<Route name="lost" path="lost/:idlost" handler={Lost} />
		<Route name="new_lost" path="lost" handler={AddLost} />
		
		<Route name="founds" path="founds" handler={Founds} />
		<Route name="found" path="found/:idfound" handler={Found} />

		<Route name="login" path="login" handler={LogIn} getUserInfo={this.getUserInfo}/>
		<Route name="signup" path="signup" handler={SignUp} />
		<DefaultRoute handler={Landing} />
	</Route>


Main component is where i put all the links to the routs. So it's like the Navigation component you defined earlier. And i have a Child LogIn Class which logs in a new user. 

My idea is that the user logged-in in the LogIn Class gets passed back to Main to display 'Connected As: '. or some other stuffs.

**Comment by tacomanator** [comment]
@ivan9074 I'm not sure exactly what part your having trouble with. Could you explain more? If you use Main as the root handler as you are, and that is where the navigation is generated, you just need to make sure that the current users' information is accessible in Main. Upon login, where do you store this information?
